lib/saml_idp/request.rb

```diff
@@ -31,48 +31,129 @@ def initialize(raw_xml = "")
       self.raw_xml = raw_xml
     end
 
+    def logout_request?
+      logout_request.nil? ? false : true
+    end
+
+    def authn_request?
+      authn_request.nil? ? false : true
+    end
+
     def request_id
-      authn_request["ID"]
+      if authn_request?
+        authn_request["ID"]
+      elsif logout_request?
+        logout_request["ID"]
+      end
+    end
+
+    def requested_authn_context
+      if authn_request? && authn_context_node
+        authn_context_node.content
+      else
+        nil
+      end
     end
 
     def acs_url
       service_provider.acs_url ||
         authn_request["AssertionConsumerServiceURL"].to_s
     end
 
+    def logout_url
+      service_provider.assertion_consumer_logout_service_url
+    end
+
+    def response_url
+      if authn_request?
+        acs_url
+      elsif logout_request?
+        logout_url
+      end
+    end
+
+    def logger(msg)
```

Comment:
logger implies we're returning some sort of log object. Maybe rename to `log` or return a logging object. The non-Rails one could delegate the function down to puts still, if you prefer to return an object